tests/sat_flows.rs

```diff
@@ -0,0 +1,308 @@
+                    amount.minus_fees
+                );
+                assert_eq!(
+                    incoming_payment_info.payment_info.payment_state,
+                    PaymentState::Succeeded
+                );
+                assert_eq!(
+                    incoming_payment_info.payment_info.invoice_details,
+                    return_invoice.clone()
+                );
+            }
+            _ => {
+                panic!("Unexpected activity: {payment:?}");
+            }
+        }
+    }
+
+    // Check whether exchange rate has updated during test run
+    let backend_exchange_rate_update_interval_secs: u64 = 5 * 60; // exchange_rate.updated_at measures the elpased time since the server updated, NOT since 3L last fetched that data.
+    let exchange_rate = sender.node.get_exchange_rate().unwrap();
+    assert!(
+        exchange_rate.updated_at.elapsed().unwrap().as_secs()
+            <= backend_exchange_rate_update_interval_secs
+    );
+}
+
+fn setup_sender_node(payment_amount_plus_fees: u64) -> TransactingNode {
+    let tn = setup_node(Some(NodeType::Sender));
+
+    let node_info = tn.node.get_node_info().unwrap();
+    let outbound_capacity = node_info.channels_info.outbound_capacity.sats;
+    assert!(
+        outbound_capacity > payment_amount_plus_fees,
+        "Sending node ({}) is insufficiently funded [Outbound capacity: {outbound_capacity}, required: {payment_amount_plus_fees}]",
+        node_info.node_pubkey
+    );
+
+    tn
+}
+
+fn setup_receiver_node(max_payment_amount: u64) -> TransactingNode {
+    let tn = setup_node(Some(NodeType::Receiver));
+
+    let node_info = tn.node.get_node_info().unwrap();
+    let inbound_capacity = node_info.channels_info.inbound_capacity.sats;
+    assert!(
+        inbound_capacity > max_payment_amount,
+        "Sending node ({}) has insufficient inbound capacity: {inbound_capacity} (required: {max_payment_amount})",
+        node_info.node_pubkey
+    );
+
+    tn
+}
+
+fn setup_node(node_type: Option<NodeType>) -> TransactingNode {
+    let received_payments = Arc::new(Mutex::new(Vec::<String>::new()));
+    let sent_payments = Arc::new(Mutex::new(Vec::<String>::new()));
+
+    let before_node_started = Instant::now();
+    let node = start_specific_node(
+        node_type.clone(),
+        Box::new(ReturnFundsEventsHandler {
+            received_payments: received_payments.clone(),
+            sent_payments: sent_payments.clone(),
+        }),
+    )
+    .unwrap();
+    println!(
+        "{:?} node started in {} milliseconds",
+        node_type.unwrap(),
+        before_node_started
+            .elapsed()
+            .as_millis()
+            .separate_with_commas()
+    );
+
+    TransactingNode {
+        node,
+        received_payments,
+        sent_payments,
+    }
+}
+
+fn wait_for_payment(
+    payment_hash: &str,
+    sender: &Arc<Mutex<Vec<String>>>,
+    receiver: &Arc<Mutex<Vec<String>>>,
```

Comment:
You could use channels here instead, but this works (just not as pretty 😃 )

Reply:
Thanks for the hint: https://github.com/getlipa/lipa-lightning-lib/pull/1190/commits/c04a8dc51f4522b2cad0391bd211db43d13cfe82